### PR TITLE
Handle pushconst reloc in elf linker

### PR DESCRIPTION
**Pull request by s-perron**
We generate a pushconst relocation for the offset of the push constant
in the spill area, but it is not handled in the linker.  This is fixed in
this commit.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: lgc/state/PipelineState.cpp [hunk 1]
```diff
@@ -540,6 +540,20 @@ void PipelineState::readUserDataNodes(Module *module) {
   m_userDataNodes = ArrayRef<ResourceNode>(m_allocUserDataNodes.get(), nextOuterNode);
 }
 
+// =====================================================================================================================
+// Find the push constant resource node.
+// Returns {topNode, node} where "node" is the found user data node, and "topNode" is the top-level user data
```
Comment (trenouf) [inline lgc/state/PipelineState.cpp:545]: Please fix the comment as it only returns a top-level node.

File: llpc/context/llpcCompiler.cpp [hunk 2]
```diff
@@ -325,7 +325,7 @@ bool VKAPI_CALL ICompiler::IsVertexFormatSupported(VkFormat format) {
 // @param options : An array of compilation-option strings
 // @param optionHash : Hash code of compilation options
 Compiler::Compiler(GfxIpVersion gfxIp, unsigned optionCount, const char *const *options, MetroHash::Hash optionHash)
-    : m_optionHash(optionHash), m_gfxIp(gfxIp) {
+    : m_optionHash(optionHash), m_gfxIp(gfxIp), m_relocatablePipelineCompilations(0) {
```
Comment (trenouf) [inline llpc/context/llpcCompiler.cpp:328]: This change seems unrelated.

File: lgc/elfLinker/ElfLinker.cpp [hunk 3]
```diff
@@ -347,14 +348,13 @@ bool ElfLinkerImpl::link(raw_pwrite_stream &outStream) {
     }
   }
 
-  // Write the PAL metadata out into the .note section.
-  writePalMetadata();
-
   // Output each section, and let it set its section table entry.
   // Ensure each section is aligned in the file by the minimum of 4 and its address alignment requirement.
```
[the remark this answers is not in the document]
Comment (s-perron) [reply lgc/elfLinker/ElfLinker.cpp:352]: Done

File: llpc/context/llpcCompiler.cpp (hunk 2 above)
Comment (s-perron) [reply llpc/context/llpcCompiler.cpp]: Move to separate PR.

File: lgc/state/PipelineState.cpp (hunk 1 above)
Comment (s-perron) [reply lgc/state/PipelineState.cpp]: Done.

File: lgc/include/lgc/state/AbiUnlinked.h [hunk 4]
```diff
@@ -75,6 +75,10 @@ const static char SamplePatternIdx[] = "$samplePatternIdx";
 // The value of the relocation is deviceIdx from the pipeline state.
 const static char DeviceIdx[] = "$deviceIdx";
 
+// Pushconst offset is "pushconst"
+// The value of the relocation is the offset of the pushconst reource node in the pipeline state.
+const static char Pushconst[] = "pushconst";
```
Comment (amdrexu) [inline lgc/include/lgc/state/AbiUnlinked.h:80]: Better to name it PushConst[] = ..
"pushconst" is accepted if you don't want "push.const" or "push_const".

File: lgc/state/PipelineState.cpp [hunk 5]
```diff
@@ -540,6 +540,17 @@ void PipelineState::readUserDataNodes(Module *module) {
   m_userDataNodes = ArrayRef<ResourceNode>(m_allocUserDataNodes.get(), nextOuterNode);
 }
 
+// =====================================================================================================================
+// Returns the resource node for the push constant.
+const ResourceNode *PipelineState::findPushconstResourceNode() const {
```
Comment (amdrexu) [inline lgc/state/PipelineState.cpp:545]: findPushconstResourceNode -> findPushConstantResourceNode

File: lgc/include/lgc/state/AbiUnlinked.h (hunk 4 above)
Comment (s-perron) [reply lgc/include/lgc/state/AbiUnlinked.h:80]: That is the name that was already in use, so I'll keep it the same.

File: lgc/state/PipelineState.cpp (hunk 5 above)
Comment (s-perron) [reply lgc/state/PipelineState.cpp]: done.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 12d256072af1c98d0bbdbdd837020839a1a37f77
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 82390674c40402c81dbf311a5ee488fb972894f5</li>
<li>SPVGEN: dev 2f679769a8491f423f5cf2f06adfa2870a82935d</li>
<li>XGL: dev 116d28a446eba3a9f8e6f577bab34fc07e090993</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev a7cfa00580cd7aa8e66f254d1e200e976d90b08e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 962/962 (100.0%)</li>
<li>Failed: 0/962 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx8</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Gfx9</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>
